### PR TITLE
Overhaul Linux file emulation

### PR DESCRIPTION
**Pull request by bradlarsen**
Previously, open file descriptors could be mapped to any of the `File`, `Directory`, `SocketDesc`, `Socket`, and `ProcSelfMaps` classes, which didn't share a common supertype, and which didn't have consistent interfaces.

Now, each of these classes implements the `FdLike` abstract base class, which specifies a number of methods.

This overhaul additionally fixes several issues with the Linux support code:

- A program that tried to ioctl on a socket would crash Manticore
- In several places, the error code returned by a failing syscall was incorrect
- Many file descriptor operations could crash Manticore when used on an already-closed descriptor
- Manticore state serialization / deserialization was incorrect for file descriptor types other than File or Socket
- In certain cases, `sys_open` / `sys_openat` could crash Manticore when trying to log a debug message
- In certain cases, `sys_readv` could crash Manticore, instead of correctly returning a negative int value
- `sys_sendfile` was broken due to a variable name typo
- In several places in Manticore’s Linux support code, incorrect type hints were present

### CODE REVIEW

File: manticore/platforms/linux.py [hunk 1]
```diff
@@ -2334,17 +2456,17 @@ def sys_send(self, sockfd, buf, count, flags):
 
         return count
 
-    def sys_sendfile(self, out_fd, in_fd, offset_p, count):
+    def sys_sendfile(self, out_fd, in_fd, offset_p, count) -> int:
         if offset_p != 0:
-            offset = self.current.read_int(offset_p, self.count.address_bit_size)
+            offset = self.current.read_int(offset_p, self.current.address_bit_size)
```
Comment (ekilmer) [inline manticore/platforms/linux.py:2461]: Nice catch!

File: manticore/platforms/linux.py [hunk 2]
```diff
@@ -2469,8 +2591,8 @@ def sched(self):
             logger.debug("SCHED:")
             logger.debug(f"\tProcess: {self.procs!r}")
             logger.debug(f"\tRunning: {self.running!r}")
-            logger.debug(f"\tRWait: {self.rwait!r}")
-            logger.debug(f"\tTWait: {self.twait!r}")
+            # logger.debug(f"\tRWait: {self.rwait!r}")
+            # logger.debug(f"\tTWait: {self.twait!r}")
```
Comment (ekilmer) [inline manticore/platforms/linux.py:2595]: I'd get rid of these since you deleted them
```suggestion
```

File: manticore/platforms/linux.py [hunk 3]
```diff
@@ -2501,14 +2624,16 @@ def wait(self, readfds, writefds, timeout):
         )
         logger.debug(f"\tProcess: {self.procs!r}")
         logger.debug(f"\tRunning: {self.running!r}")
-        logger.debug(f"\tRWait: {self.rwait!r}")
-        logger.debug(f"\tTWait: {self.twait!r}")
+        # logger.debug(f"\tRWait: {self.rwait!r}")
+        # logger.debug(f"\tTWait: {self.twait!r}")
```
Comment (ekilmer) [inline manticore/platforms/linux.py:2628]:
```suggestion
```

**Comment by ekilmer** [comment]
This PR is looking really good so far!! Thank you for taking the time to refactor this!

File: manticore/platforms/linux.py [hunk 4]
```diff
@@ -74,7 +77,127 @@ def mode_from_flags(file_flags: int) -> str:
...
+class FdTable:
+    """
+    This represents Linux's file descriptor table.
+
+    Each file descriptor maps to an C{FdLike} object.  Additionally, each file
+    descriptor maps to a set of PIDs of processes that are waiting on that
+    descriptor.
+    """
+
+    __slots__ = ["_mapping", "_closed_fds"]
```
[the remark this answers is not in the document]
Comment (bradlarsen) [reply manticore/platforms/linux.py:130]: Good catch, thanks!  That was a leftover.

File: manticore/platforms/linux.py [hunk 5]
```diff
@@ -673,8 +816,8 @@ def _init_std_fds(self) -> None:
     def _init_cpu(self, arch: str) -> None:
         # create memory and CPU
         cpu = self._mk_proc(arch)
-        self.procs = [cpu]
-        self._current = 0
+        self.procs: List[Cpu] = [cpu]
+        self._current: Optional[int] = 0
```
[the remark this answers is not in the document]
Comment (bradlarsen) [reply manticore/platforms/linux.py:820]: Yeah, it _is_ interesting that `self._current` can be `None`.  mypy actually helped point out that this is the case... particularly in existing Linux code related to scheduling.

File: manticore/platforms/linux.py (hunk 1 above)
Comment (bradlarsen) [reply manticore/platforms/linux.py:2461]: Thanks :)  Adding the type hints there got mypy looking at the implementation, and it complained that `self.count` was nonsense.

There are actually a few dozen places elsewhere in Manticore where there are issues like that, or use-of-undefined variables.  I'm gonna open another PR for those sometime soon.

**Comment by bradlarsen** [comment]
Thanks for the review, @ekilmer :)  It seems like I still have a bit of work to do on this PR, including adding some tests.

**Comment by ekilmer** [comment]
Finally got around to testing this on my branch. Seems to be working!! Great changes!

File: manticore/platforms/linux.py [hunk 6]
```diff
@@ -436,10 +578,10 @@ def connect(self, peer):
         if peer.peer is None:
             peer.peer = self
 
-    def read(self, size):
+    def read(self, size: int = -1):
```
[the remark this answers is not in the document]
Comment (ekilmer) [reply manticore/platforms/linux.py:620]: I think this commit https://github.com/trailofbits/manticore/pull/1673/commits/27414ec74714cbfdccdbfd6f5427566b9a8987fa missed this spot ;)

[the remark this answers is not in the document]
Comment (ekilmer) [reply manticore/platforms/linux.py:620]:
```suggestion
    def read(self, size: int):
```

**Comment by bradlarsen** [comment]
> Finally got around to testing this on my branch. Seems to be working!! Great changes!

Terrific, thanks @ekilmer for the test!

**Pull request merged**